### ISSUE TITLE
Support using a PipelineTask in the Pipeline CRD to run other Pipelines the same way we run a Task (nest pipelines in pipelines)

### ISSUE THREAD

**Issue by skaegi** [tektoncd/pipeline#2134]
Pipelines group tasks together in a workflow. It would be incredibly useful to be able to re-use that mechanism to create sub-workflows. 

Pipelines and Tasks both take Params, Workspaces, and generate Results so with a little work it should be possible to use a PipelineRef in place of a TaskRef when defining a Pipeline Task. For the other aspects required to generate a "sub"-PipelineRun these could be inherited from the parent PipelineRun.

Here's a basic example to illustrate syntax...
```
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: pipeline-calling-pipeline
spec:
  params:
  - name: best-param-ever
  workspaces:
  - name: pipeline-ws1
  tasks:
  - name: pipeline-in-task-clothing
    pipelineRef:
      name: some-sub-pipeline
    workspaces:
    - name: sub-pipeline-ws0
      workspace: pipeline-ws1
  - name: regular-task
    taskRef:
      name: some-task
    params:
    - name: new-input
      value: $(tasks.pipeline-in-task-clothing.results.some-ouput)
    workspaces:
    - name: src
      workspace: pipeline-ws1
    runAfter:
    - pipeline-in-task-clothing
```

In a nutshell, other than using pipelineRef in-place of taskRef the definitions behave the same.

**Comment by bobcatfish** [comment]
> pipeline-in-task-clothing

![image](https://user-images.githubusercontent.com/432502/79778454-c1a3c780-8306-11ea-998c-27c0c4ed059f.png)

**Comment by jlpettersson** [comment]
>Pipelines and Tasks both take Params, Workspaces, and generate Results so with a little work it should be possible to use a PipelineRef in place of a TaskRef when defining a Pipeline Task. For the other aspects required to generate a "sub"-PipelineRun these could be inherited from the parent PipelineRun.

I feel the use case is a bit unclear here. But I do have things that reminds me about this. E.g. The use-case I describe in https://github.com/tektoncd/pipeline/issues/2505#issue-608585765 - there I use (or "link") two `PipelineRun` and reuse the `workspace` volume.

On the other hand, I will have another use-case that also reminds a bit. But one fundamental difference is that I probably want to have it _asynchronous_. When talking about "linking" `Pipelines` or "chain", it may be done _synchronous_ or _asynchronous_. That is also related to e.g. introducing _asynchronous_ tasks - but I need to think about it more.

**Comment by madbence** [comment]
This feature would be really useful for us. eg. right now we have a task that can set the status on a given commit, and a task that can execute unit tests (eg. run `npm test`, etc.). We'd like to distribute a task/pipeline that can combine these (eg. set the commit to pending, run unit tests, set commit status to success/failure).

![nested-pipeline](https://user-images.githubusercontent.com/296735/82465905-11211300-9ac0-11ea-9d6d-4fe9475c46a9.png)

we'd like to provide to option to use `set-status` and `run-unit-tests` in a pipeline, and to be able to use a "higher-level" task, like "run-unit-tests-with-commit-status".

**Comment by bobcatfish** [comment]
Thanks for the use case @madbence this looks exactly along the lines we're thinking as well - just to double check, set-status sets the status of something like a GitHub check (or similar using non-GitHub)?

**Comment by bobcatfish** [comment]
oh and i forgot to post here some relevant links:

- [WIP design doc for pipelines in pipelines](https://docs.google.com/document/d/14Uf7XQEnkMFBpNYRZiwo4dwRfW6do--m3yPhXHx4ybk/edit) - visible to [tekton-dev](https://github.com/tektoncd/community/blob/master/contact.md#mailing-list)
- Also @ImJasonH has been looking into some thoughts around #215 which may let us conquer 2 issues with one stone

**Comment by madbence** [comment]
@bobcatfish exactly, `set-status` creates a [commit status](https://developer.github.com/v3/repos/statuses/#create-a-status).

**Comment by vdemeester** [comment]
cc @mattmoor

**Comment by bobcatfish** [comment]
[Tekton custom tasks](https://docs.google.com/document/d/10nQSeIse7Ld4fLg4lhfgUmNKtewfaFNET3zlMdRnBuQ/edit) would let us have this feature and more; planning to put this feature on hold until we have custom tasks and then we can decide if we want it as a top level feature or not.

**Comment by dbazhal** [comment]
Couldn't this issue be done without custom tasks? We're in pain from disability to include pipelines into pipelines :) Problem is that we have to include task "run_pipeline" each time when we want to include pipeline inside pipeline. Each time such "incuded" pipeline fails, there is no easy way to find out why subpipeline failed, or what tasks it contained, etc.

**Comment by JPEWdev** [comment]
I also would really like support for running pipelines inside of pipelines. This is really useful for the case were you want some "super pipeline" that runs different variants of the same sub pipeline with different arguments (i.e. like a matrix/parameterized build from most other CI systems), provided that it can run all the sub-pipelines in parallel with different workspaces.

**Comment by JPEWdev** [comment]
> I also would really like support for running pipelines inside of pipelines. This is really useful for the case were you want some "super pipeline" that runs different variants of the same sub pipeline with different arguments (i.e. like a matrix/parameterized build from most other CI systems), provided that it can run all the sub-pipelines in parallel with different workspaces.

Also I would like to point out that it would only *really* be useful if the subpipelines could be scheduled on different nodes. My (limited) understanding of tekton is that this might be difficult if the parent pipeline has to specify the workspace

**Comment by dbazhal** [comment]
> Also I would like to point out that it would only _really_ be useful if the subpipelines could be scheduled on different nodes. My (limited) understanding of tekton is that this might be difficult if the parent pipeline has to specify the workspace

agreed. Work with workspaces is sometimes pain, double pain if super-pipeline can't create ws, so Super-pipeline runs 1) task to create ws 2) each sub-pipeline with passing it workspace as argument 3) has to finish with task that deleted ws.

Also, i dont know how it could be done, but it would be better to aggregate sub-pipelines bodies into single pipelinerun. Creating separate pipeline run for each "sub"-pipeline makes it difficult to read logs of each task in each sub-pipeline(even more difficult with openshift tekton pipelines web-interface), it makes sub-pipelines execute slower(running pipelinerun isnt flash-fast itself, and it makes sometimes passing same arguments to each sub-pipeline a lot of copy-paste work.

**Comment by bobcatfish** [comment]
Quick update from my side: I am (at least!) very strongly in favor of this feature: I think it solves a lot of problems we have around expressing more complex error handling without adding a lot of complication to our API.

> Couldn't this issue be done without custom tasks?

Theoretically it could - the coupling to custom tasks is more about the fact that custom tasks make this really easy to prototype; and make it so folks can start using Pipelines within Pipelines right away without us explicitly needing to add support upstream.

We basically just need someone to have the time to prototype + propose + drive this forward. I haven't had the time so I'll unassign myself for now.

**Comment by HerrmannHinz** [comment]
super helpful. would finally allow us to "chain" pipelines. plz 🙏

**Comment by aelbarkani** [comment]
+1 for this feature, very helpful for reusing pipelines !

**Comment by jerop** [comment]
TEP-0056 proposing Pipelines in Pipelines is _implementable_ - we welcome contributions in implementing it 

https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md

/help-wanted

**Comment by EmmaMunley** [comment]
Happy to start working on implementing TEP-0056.

/assign @EmmaMunley

**Comment by ChrisPaul33** [comment]
Hi, there are any updates about the implementation of the feature?

**Comment by pritidesai** [comment]
Hi @ChrisPaul33, we have started implementing this feature. What API version are you currently using? Looking for some insight for - https://github.com/tektoncd/pipeline/pull/7055#discussion_r1298820844

**Comment by ChrisPaul33** [comment]
Hello @pritidesai, we are currently using v1beta1.
We are still trying to migrate to Tekton so migrating from v1beta1 to v1 for us won't be a big deal.
I will try my best to look at #7055.

**Comment by vdemeester** [comment]
@twoGiants @waveywaves @afrittoli I guess this can now be closed (Pipelines in Pipelines is there, in alpha, but there)
